File: server/relay_server.js
const WebSocket = require("ws");

const wss = new WebSocket.Server({ port: 6789 });
const agents = new Map();

console.log("🚀 WebSocket 服务器已启动，监听端口 6789");

wss.on("connection", (ws) => {
  ws.once("message", (message) => {
    const msg = message.toString();

    // ✅ 控制端请求在线设备列表
    if (msg === "controller:list") {
      const onlineAgents = Array.from(agents.keys());
      if (onlineAgents.length === 0) {
        ws.send("NO_AGENT");
      } else {
        ws.send(onlineAgents.join(","));
      }
      ws.close(); // 关闭连接，因为这个只用于获取列表
      return;
    }

    // ✅ 被控端连接
    if (msg.startsWith("agent:")) {
      const code = msg.split(":")[1];
      agents.set(code, ws);
      console.log(`✅ Agent [${code}] connected`);

      ws.on("close", () => {
        agents.delete(code);
        console.log(`🧹 Agent [${code}] disconnected`);
      });
    }

    // ✅ 控制端连接并指定控制某个 agent
    else if (msg.startsWith("controller:")) {
      const code = msg.split(":")[1];
      const agentWs = agents.get(code);
      console.log(`🎮 Controller requests access to Agent [${code}]`);

      if (!agentWs || agentWs.readyState !== WebSocket.OPEN) {
        ws.send("NO_AGENT");
        ws.close();
        return;
      }

      ws.send("CONNECTED");

      ws.on("message", (cmd) => {
        try {
          agentWs.send(cmd.toString());
          agentWs.once("message", (data) => {
            ws.send(data.toString());
          });
        } catch (err) {
          ws.send("ERROR");
        }
      });
    }
  });
});

File: server/relay_server.py
import asyncio
import websockets

clients = {}

async def relay(websocket, path):
    try:
        client_type = await websocket.recv()
        if client_type.startswith("agent:"):
            code = client_type.split(":")[1]
            clients[code] = websocket
            print(f"✅ Agent [{code}] connected")
            try:
                while True:
                    await asyncio.sleep(10)
            except Exception as e:
                print(f"❌ Agent [{code}] 中断: {e}")
            finally:
                if clients.get(code) == websocket:
                    del clients[code]
                    print(f"🧹 Agent [{code}] 清理完成")
        elif client_type.startswith("controller:"):
            code = client_type.split(":")[1]
            agent_ws = clients.get(code)
            print(f"🎮 控制端请求连接 Agent [{code}]")
            if agent_ws:
                await websocket.send("CONNECTED")
                try:
                    while True:
                        cmd = await websocket.recv()
                        print(f"📤 控制命令: {cmd}")
                        await agent_ws.send(cmd)
                        resp = await agent_ws.recv()
                        await websocket.send(resp)
                except Exception as e:
                    print(f"❌ 控制端异常：{e}")
            else:
                await websocket.send("NO_AGENT")
    except Exception as e:
        print(f"🔥 服务器异常：{e}")

async def main():
    async with websockets.serve(relay, "0.0.0.0", 6789):
        print("🚀 WebSocket 服务器已启动，监听端口 6789")
        await asyncio.Future()

asyncio.run(main())
